sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Reshuffle.java

```diff
@@ -94,4 +105,31 @@ public void processElement(ProcessContext c) {
             "RestoreOriginalTimestamps",
             ReifyTimestamps.<K, V>extractFromValues());
   }
+
+  /** Implementation of {@link #viaRandomKey()}. */
+  public static class ViaRandomKey<T> extends PTransform<PCollection<T>, PCollection<T>> {
+    private ViaRandomKey() {}
+
+    @Override
+    public PCollection<T> expand(PCollection<T> input) {
+      return input.apply(
+          "Pair with random key",
+          ParDo.of(
+              new DoFn<T, KV<Integer, T>>() {
+                private int shard;
+
+                @Setup
+                public void setup() {
+                  shard = ThreadLocalRandom.current().nextInt();
+                }
+
+                @ProcessElement
+                public void processElement(ProcessContext context) {
+                  context.output(KV.of(++shard, context.element()));
```

Comment:
Wondering if there are unexpected consequences of not using random key each time. E.g. lets say for some reason, every 10th element is very large and take a long time to process. All those elements could end up on one or small subset of workers.

Reply:
Not sure what you mean? This key has exactly the same statistical properties as calling ThreadLocalRandom.current().nextInt() for every element.

Reply:
Assuming runner selects range/worker using 'key.hashCode() % num_ranges', then every nth element goes to the same range?

Reply:
Key is a growing sequence k0, k0+1, k0+2, ... where k0 is random - but hash(k0+n) != hash(k0) + n.

Reply:
Per offline discussion: good point, some runners may be naive about hashing the key; fixed this.

Reply:
Very nice fix. Thanks.